fvm/state/state.go

```diff
@@ -79,132 +84,210 @@ func WithMaxInteractionSizeAllowed(limit uint64) func(st *State) *State {
+func (s *State) Get(owner, controller, key string) (flow.RegisterValue, error) {
 	if err := s.checkSize(owner, controller, key, []byte{}); err != nil {
 		return nil, err
 	}
 
-	// check draft first
-	if p, ok := s.draft[fullKey(owner, controller, key)]; ok {
-		// just call the ledger get for tracking touches
-		_, err := s.ledger.Get(owner, controller, key)
-		if err != nil {
-			return nil, &LedgerFailure{err}
-		}
-		return p.value, nil
+	pKey := PayloadKey{owner, controller, key}
+	s.LogTouch(&Payload{pKey, nil})
+
+	value, err := s.get(pKey)
+	if err != nil {
+		return nil, err
 	}
 
-	// return from draft
-	if p, ok := s.draft[fullKey(owner, controller, key)]; ok {
-		// just call the ledger get for tracking touches
-		_, err := s.ledger.Get(owner, controller, key)
-		if err != nil {
-			return nil, &LedgerFailure{err}
-		}
-		return p.value, nil
+	// update read catch
+	p := Payload{pKey, value}
+	s.readCache[pKey] = p
+	s.ReadCounter++
+	s.TotalBytesRead += p.size()
+	return value, s.checkMaxInteraction()
+}
+
+func (s *State) GetWithoutTracking(owner, controller, key string) (flow.RegisterValue, error) {
+	pKey := PayloadKey{owner, controller, key}
+	value, err := s.get(pKey)
+	if err != nil {
+		return nil, err
+	}
+	p := Payload{pKey, value}
+	s.readCache[pKey] = p
+	return value, nil
+}
+
+// Get returns a register value given owner, controller and key
+func (s *State) get(pKey PayloadKey) (flow.RegisterValue, error) {
+	// check delta first
+	if p, ok := s.delta[pKey]; ok {
+		return p.Value, nil
 	}
 
 	// return from read cache
-	if p, ok := s.readCache[fullKey(owner, controller, key)]; ok {
-		// just call the ledger get for tracking touches
-		_, err := s.ledger.Get(owner, controller, key)
-		if err != nil {
-			return nil, &LedgerFailure{err}
-		}
-		return p.value, nil
+	if p, ok := s.readCache[pKey]; ok {
+		return p.Value, nil
+	}
+
+	// read from parent
+	if s.parent != nil {
+		value, err := s.parent.GetWithoutTracking(pKey.Owner, pKey.Controller, pKey.Key)
+		return value, err
 	}
 
 	// read from ledger
-	value, err := s.ledger.Get(owner, controller, key)
+	value, err := s.ledger.Get(pKey.Owner, pKey.Controller, pKey.Key)
 	if err != nil {
 		return nil, &LedgerFailure{err}
 	}
+	return value, nil
+}
 
-	// update read catch
-	s.readCache[fullKey(owner, controller, key)] = payload{owner, controller, key, value}
+func (s *State) updateDelta(p *Payload) {
+	// check if a delta already exist for this key
+	// reduce the bytes to be written
+	if old, ok := s.delta[p.PayloadKey]; ok {
+		s.ToBeWrittenCounter--
+		s.TotalBytesToBeWritten -= old.size()
+	}
 
-	return value, s.updateInteraction(owner, controller, key, value, []byte{})
+	s.delta[p.PayloadKey] = *p
+	s.ToBeWrittenCounter++
+	s.TotalBytesToBeWritten += p.size()
 }
 
+// Set updates state delta with a register update
 func (s *State) Set(owner, controller, key string, value flow.RegisterValue) error {
```

Comment:
I expected the write counter to be incremented here... But now I'm not sure how the write counter works.

As a separate question: should the write counter be incremented twice if we are writhing twice to the same key.